### ISSUE TITLE
cram: add textual note to indicate semantics of ReadItem

### ISSUE THREAD

**Issue by kortschak** [samtools/hts-specs#466]
In the CRAM spec, the pseudo-code makes reference to a `ReadItem`, which appears to take two parameters, a name and a type. The values though are defined by place rather than an actual label, so this parameter seems more like a decoration than an actual parameter.

It would be nice to have an unambiguous explanation of the intended semantics here, as there is for the rANS decode pseudocode. Though there, the explanatory text is at odds with one of the uses; "The function `ReadByte` below is undefined, but is expected to fetch the next single unsigned byte from an unspecified input source," but in `RansDecode`, there is a specified input. Maybe specify the input parameter universally and change the text, or remove the input parameter from `RansDecode`.

/cc @jkbonfield

**Comment by jkbonfield** [comment]
Apologies for forgetting about this one.

The pseudocode does indeed play a bit fast and loose at times with regards to parameters and return values.  It's intended to be understandable and clear to humans, if not explicitly to a computer.  However in this case it's evidently failing in that goal.

I'm not sure I understand the question.  Do you mean you'd prefer `ReadItem(Series=BF, Type=INT)` over `ReadItem(BF, INT)`?  The pseudocode is deliberately as devoid of syntactic sugar as it can be.  ReadItem isn't ever defined anywhere.  It should have some introductory text as with ReadByte explaining the intended purpose, which can then explain it takes two arguments: a dataseries name and a data type.

I was much more thorough when I wrote the CramCodecs document for the CRAM 3.1 prototype.  The PR can be seen in https://github.com/samtools/hts-specs/pull/433
and a PDF copy is here:  [CRAMcodecs.pdf](https://github.com/samtools/hts-specs/files/4954044/CRAMcodecs.pdf)


The second issue you raise with `RansDecode` is a trickier one.  Generally I tried to avoid tedious things like defining data sources as we don't need to know if we're reading from a file, a pipe, stdin, or whatever.  It's just a magic "get me the next bytes".  Unfortunately that works everywhere except for one place in CRAM 3.1 (not official, but I was probably thinking of it while doing this), which is the nested nature of the newer rANS where the order-1 frequency table can itself be compressed using the order-0 rANS codec and hence the source changes.  Again the desire to not explicitly list input sources is purely for brevity as that, generally, makes for more human readable pseudocode.

Note rANS was the first thing I wrote any pseudocode for so there's probably been some change of style.  I'm not sure if it's appropriate.  I'm inclined to either drop the input parameter of `RansDecode` or add a comment indicating that any byte/int decoding that does not specify the input will be using the supplied `input` datastream.  However ultimately the rANS pseudocode in the CRAMv3.tex document will completely vanish once the cram codecs PR is merged, so it's best to review that for changes in that specific area.
